File: cpp/0096.0_Unique_Binary_Search_Trees.cpp
/*
DP

Runtime: 0 ms, faster than 100.00% of C++ online submissions for Unique Binary Search Trees.
Memory Usage: 6 MB, less than 64.16% of C++ online submissions for Unique Binary Search Trees.
*/
class Solution {
public:
    int numTrees(int n) {
        vector<int> memo(n + 1, 1);
        int nodes = 2;
        while (nodes <= n) {
            int count = 0;
            for (int i = 1; i <= nodes; i++) {
                int leftNodes = i - 1;
                int rightNodes = nodes - i;
                count += memo[leftNodes] * memo[rightNodes];
            }
            memo[nodes] = count;
            nodes++;
        }
        return memo[n];
    }
};

File: java/0096.0_Unique_Binary_Search_Trees.java
/*
DP
T: O(N^2)
S: O(N)

Runtime: 0 ms, faster than 100.00% of Java online submissions for Unique Binary Search Trees.
Memory Usage: 35.5 MB, less than 93.29% of Java online submissions for Unique Binary Search Trees.
*/

class Solution {
    public int numTrees(int n) {
        int[] G = new int[n + 1];
        G[0] = G[1] = 1;
        for (int i = 2; i <= n; i++) {
            for (int j = 1; j <= i; j++) {
                G[i] += G[j - 1] * G[i - j];
            }
        }
        return G[n];
    }
}
